### ISSUE TITLE
Add BrowserStack Support

### ISSUE THREAD

**Issue by yanatan16** [defunctzombie/zuul#124]
[BrowserStack](http://browserstack.com) is a better manual testing product than sauce labs. Since we're planning on stopping paying for Sauce, adding automated browser testing with BrowserStack is a good idea. I'll probably get a start on this in a bit.

**Comment by defunctzombie** [comment]
There were some previous PRs about this but I ran into issues with BrowserStack queuing. These PRs (now closed) are likely outdated but might be a good starting point. I will happily review a new PR.

**Comment by yanatan16** [comment]
I ran into the same(?) queueing bug. When multiple browsers are used, it doesn't queue right and starts yelling at you about # of concurrent tests. I wonder if its fixable.

**Comment by defunctzombie** [comment]
I tried to tweet at them to explain how it is something they should fix but
they ignored me.

I think that is why I gave up initially.

Might have to detect this in zuul and just do your own queue and retry.

**Comment by yanatan16** [comment]
Yeah it looks like in their own examples (https://github.com/browserstack/selenium-runner) they manually setup concurrency using `async.queue`.

**Comment by defunctzombie** [comment]
Oh I remember one nasty thing that made the queue harder to implement and
why it is something I felt they needed to fix.

If you have two separate Zuuls running, they will think that have access to
some N concurrency, but in reality the concurrency is for your entire
account (not session) iirc. So they will still fail out of the queue. You
have to constantly check for avail slots and even then there could be a
race so you gotta handle that too. Basically gotta have lots of retry logic
and detect concurrency errors.

**Comment by andyhu** [comment]
There is an article shows how to use browserstack without even using zuul:
https://joshduff.com/2015-08-automated-testing-with-tape-and-browserstack.md
